### PR TITLE
Quick Reblog: Enable reblogging posts into communities

### PR DESCRIPTION
**Pull request by marcustyphoon**
### Description

Adds communities the user has joined to the Quick Reblog blog selector dropdown.

Note that posts drafted or queued into a community (presumably) disappear into the ether, so this puts up an error modal if the user tries to do so. (Ideally it should probably grey out the buttons in question; I was hoping one could css select based on a `<select>`'s current value but I'm not sure you can, so this would probably have to be a bunch of js.)

Should alreadyReblogged exclude reblogging into a community? ¯\\\_(ツ)\_/¯ 

### Testing steps

At time of writing, I have a spare community with invite link [tumblr.com/join/oki7cQsr](https://www.tumblr.com/join/oki7cQsr).

todo: list performed tests

### CODE REVIEW

File: src/features/quick_reblog.js [hunk 1]
```diff
@@ -331,12 +334,24 @@ export const main = async function () {
     alreadyRebloggedLimit
   } = await getPreferences('quick_reblog'));
 
+  joinedCommunities = await apiFetch('/v2/communities')
+    .then(({ response }) => response)
+    .catch(() => []);
```
Comment (AprilSylph) [inline src/features/quick_reblog.js:339]: This should probably be exported by `src/utils/user.js`, no? It doesn't come from `/v2/user/info` but it _is_ basic user-specific data that we might need for other features in the future.

---

feature future
future feature
future feature future
feature future feature

Comment (marcustyphoon) [reply src/features/quick_reblog.js:339]: Oh yeah, probably!

File: src/utils/user.js [hunk 2]
```diff
@@ -39,3 +44,8 @@ export const adminBlogs = userInfo?.blogs?.filter(blog => blog.admin) ?? [];
  * {string[]} adminBlogNames - An array of blog names the current user is admin of
  */
 export const adminBlogNames = adminBlogs.map(blog => blog.name);
+
+/**
+ * {object?} userInfo - The contents of the /v2/user/info API endpoint
+ */
+export const joinedCommunities = fetchedCommunitiesInfo.response;
```
Comment (AprilSylph) [inline src/utils/user.js:51]: :coppy:

Comment (marcustyphoon) [reply src/utils/user.js:51]: er.

Comment (marcustyphoon) [reply src/utils/user.js]: not sure how I feel about AI suggestions but I wouldn't mind a commit hook with a little language model that just piped up and said "hey this is probably wrong"

File: src/features/quick_reblog.js [hunk 3]
```diff
@@ -67,17 +67,15 @@ const alreadyRebloggedStorageKey = 'quick_reblog.alreadyRebloggedList';
 const rememberedBlogStorageKey = 'quick_reblog.rememberedBlogs';
 const quickTagsStorageKey = 'quick_tags.preferences.tagBundles';
 const blogHashes = {};
+const avatarUrls = {};
```
Comment (AprilSylph) [inline src/features/quick_reblog.js:70]: When we're not setting our own key-values as part of the definition, I think maps should be actual `Map`s. Empty `Record`s are spooky, unnecessary square bracket notation is spooky.

(Sin committed: criticise someone for following an established pattern in a file!)

Bonus points if you convert `blogHashes` to a `Map` as well. <3

Comment (marcustyphoon) [reply src/features/quick_reblog.js:70]:
```diff
- const savedBlogUuid = Object.keys(blogHashes).find(uuid => blogHashes[uuid] === savedBlogHash);
+ const savedBlogUuid = [...blogHashes].find(([uuid, hash]) => hash === savedBlogHash)?.[0];
```

```diff
- const savedBlogUuid = Object.keys(blogHashes).find(uuid => blogHashes[uuid] === savedBlogHash);
+ const savedBlogUuid = [...blogHashes.keys()].find(uuid => blogHashes.get(uuid) === savedBlogHash);
```

hm

Comment (marcustyphoon) [reply src/features/quick_reblog.js]: That's interesting! I'm curious if you had that stance before (I assume) doing a lot of typescript. That was kind of my experience.

**Review by AprilSylph — CHANGES_REQUESTED** [review]
I think it's worth doing the JS to hide the Queue and Draft buttons in the popup entirely when a community is selected.

The good news is that the CSS isn't complicated because I'm awesome, it just needs to be something like `.actionButtons.communitySelected button:not([data-state="published"]) { display: none; }`. The bad news is that you'll need another `change` event listener on `blogSelector` that looks up whether the new value is from `joinedCommunities` or not to add or remove the necessary class from `actionButtons`.

...which I guess isn't too bad because you could probably make a `Set` of just the user's joined community UUIDs for easier lookup?
Maybe even define and export that in `utils/user.js`.
Maybe even make an equivalent for the user's blogs too, I'm sure we could make use of that.

It is 10pm here. I'm logging off before I develop crippling insomnia again.

**Comment by marcustyphoon** [comment]
> ...which I guess isn't too bad because you could probably make a Set of just the user's joined community UUIDs for easier lookup?

I couldn't think of an advantage or disadvantage of Array.includes vs Set.has, really. Just put in what was consistent with the rest of the util, for now, but I could see it making sense to make any export of a bunch of unique strings a set. If we're never going to be doing union/intersection/etc with these it's just a matter of taste, I assume.

**Review by AprilSylph — APPROVED** [review]
:heart:

**Pull request merged**